cmd/collector/app/zipkin/http_handler_test.go

```diff
@@ -0,0 +1,283 @@
+	span := &zipkincore.Span{}
+	spans := []*zipkincore.Span{}
+	spans = append(spans, span)
+	server, _ := initializeTestServer(nil)
+	defer server.Close()
+
+	statusCode, _, err := postBytes(server.URL+`/api/v1/spans`, []byte{}, createHeader("application/json"))
+	assert.NoError(t, err)
+	assert.EqualValues(t, http.StatusBadRequest, statusCode)
+}
+
+func zipkinSerialize(spans []*zipkincore.Span) []byte {
+	t := thrift.NewTMemoryBuffer()
+	p := thrift.NewTBinaryProtocolTransport(t)
+	p.WriteListBegin(thrift.STRUCT, len(spans))
+	for _, s := range spans {
+		s.Write(p)
+	}
+	p.WriteListEnd()
+	return t.Buffer.Bytes()
+}
+
+func TestFormatBadBody(t *testing.T) {
+	server, _ := initializeTestServer(nil)
+	defer server.Close()
+	bodyBytes := []byte("not good")
+	statusCode, resBodyStr, err := postBytes(server.URL+`/api/v1/spans`, bodyBytes, createHeader("application/x-thrift"))
+	assert.NoError(t, err)
+	assert.EqualValues(t, http.StatusBadRequest, statusCode)
+	assert.EqualValues(t, "Unable to process request body: *zipkincore.Span field 0 read error: EOF\n", resBodyStr)
+}
+
+func TestDeserializeZipkinWithBadListStart(t *testing.T) {
+	span := &zipkincore.Span{TraceID: 12, Name: "test"}
+	spans := []*zipkincore.Span{}
+	spans = append(spans, span)
+	spanBytes := zipkinSerialize(spans)
+	_, err := deserializeZipkin(append([]byte{0, 255, 255}, spanBytes...))
+	assert.Error(t, err)
+}
+
+func TestCannotReadBodyFromRequest(t *testing.T) {
+	handler := NewAPIHandler(&mockZipkinHandler{})
+	req, err := http.NewRequest(http.MethodPost, "whatever", &errReader{})
+	assert.NoError(t, err)
+	rw := dummyResponseWriter{}
+	handler.saveSpans(&rw, req)
+	assert.EqualValues(t, http.StatusInternalServerError, rw.myStatusCode)
+	assert.EqualValues(t, "Unable to process request body: Simulated error reading body\n", rw.myBody)
+}
+
+type errReader struct{}
+
+func (e *errReader) Read(p []byte) (int, error) {
+	return 0, fmt.Errorf("Simulated error reading body")
+}
+
+type dummyResponseWriter struct {
+	myBody       string
+	myStatusCode int
+}
+
+func (d *dummyResponseWriter) Header() http.Header {
+	return http.Header{}
+}
+
+func (d *dummyResponseWriter) Write(bodyBytes []byte) (int, error) {
+	d.myBody = string(bodyBytes)
+	return 0, nil
+}
+
+func (d *dummyResponseWriter) WriteHeader(statusCode int) {
+	d.myStatusCode = statusCode
+}
+
+func createHeader(contentType string) *http.Header {
+	header := &http.Header{}
+	if len(contentType) > 0 {
+		header.Add("Content-Type", contentType)
+	}
+	return header
+}
+
+func gzipEncode(b []byte) []byte {
+	buffer := &bytes.Buffer{}
+	z := gzip.NewWriter(buffer)
+	z.Write(b)
+	z.Close()
+	return buffer.Bytes()
+}
+
+func postBytes(urlStr string, bytesBody []byte, header *http.Header) (int, string, error) {
+	req, err := http.NewRequest(http.MethodPost, urlStr, bytes.NewBuffer(bytesBody))
+	if err != nil {
+		return 0, "", err
+	}
+	if strings.Contains(http.DetectContentType(bytesBody), "gzip") {
```

Comment:
why not explicitly pass this header in when you're testing gzip? Feels weird(?) that the test util is adding the header for you after detecting it's gzip.